File: 1семестр/6/6(4)/6(4)/stack.cpp
#include "stack.h"

struct ElementOfStack
{
	TypeOfElement element;
	ElementOfStack *next;
};

struct Stack
{
	ElementOfStack *head;
	int size;
};

Stack *createStack()
{
	Stack *newStack = new Stack;
	newStack->size = 0;
	newStack->head = nullptr;
	return newStack;
}

void push(Stack *stack, TypeOfElement element)
{
	ElementOfStack *newElement = new ElementOfStack;
	newElement->element = element;
	newElement->next = stack->head;
	stack->head = newElement;
	stack->size++;
}

void pop(Stack *stack)
{
	ElementOfStack *temp = stack->head;
	stack->head = stack->head->next;
	delete temp;
	stack->size--;
}

TypeOfElement top(Stack *stack)
{
	return stack->head->element;
}

int size(Stack *stack)
{
	return stack->size;
}

void deleteStack(Stack *stack)
{
	ElementOfStack *temp = stack->head;
	while (temp != nullptr)
	{
		ElementOfStack *tempForDelete = temp;
		temp = temp->next;
		delete tempForDelete;
	}
	delete stack;
}

File: 2семестр/2/Stack/Stack/Stack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Stack
{
    /// <summary>
    /// Стэк
    /// </summary>
    /// <typeparam name="T">тип,который нам нужен</typeparam>
    public class Stack<T>
    {
        private ElementOfStack head;
        private int size;
        private class ElementOfStack
        {
            public T Value { get; set; }
            public ElementOfStack Next { get; set; }
            public ElementOfStack(Stack<T> stack)
            {
                this.Next = stack.head;
            }
        }
        public Stack()
        {
            head = null;
            size = 0;
        }

        /// <summary>
        /// Метод, который вставляет элемент в "голову"
        /// </summary>
        /// <param name="value">элемент, который мы хотим поместить в "голову"</param>
        public void Push(T value)
        {
            ElementOfStack newElement = new ElementOfStack(this);
            newElement.Value = value;
            head = newElement;
            size++;
        }

        /// <summary>
        /// Удаляет элемент в "голове"
        /// </summary>
        public void Pop()
        {
            if (head != null)
            {
                this.head = this.head.Next;
                size--;
            }
            else
            {
                throw new Exception();
            }
        }

        /// <summary>
        /// Возвращает размер стэка
        /// </summary>
        /// <returns>размер стэка</returns>
        public int Size()
        {
            int count = 0;
            if (head != null)
            {
                while (this.head != null)
                {
                    this.head = this.head.Next;
                    count++;
                }
                return count;
            }
            else
            {
                return count;   
            }
        }

        /// <summary>
        /// ВОзвращает элемент в "голове"
        /// </summary>
        /// <returns>элемент в "голове"</returns>
        public T Top()
        {
            if (head != null)
            {
                return head.Value;
            }
            else
            {
                throw new Exception();
            }
        }

        /// <summary>
        /// Обнуляет стэк
        /// </summary>
        public void Clear()
        {
            head = null;
            size = 0;
        }

        /// <summary>
        /// Печатает стэк
        /// </summary>
        /// <returns>элементы стэка</returns>
        public string Write()
        {
            string result = "";
            for (ElementOfStack temp = head; temp != null; temp = temp.Next)
            {
                result += temp.Value;
                result += " ";
            }
            return result;
        }
    }
}
